### ISSUE TITLE
headers.inc.php library calling refactor needed

### ISSUE THREAD

**Issue by MigDinny** [LibreHealthIO/LibreEHR#915]
I realized that we got a function _call_required_libraries_ that a lot of files don't actually use.

Some files actually call some libraries/scripts twice unnecessarily. 

demographics.php is an example I'm working on at the moment that has that kind of problems.

I could work on a refactor of this kind on the remaining of the project. I'll submit a PR to show what I actually did, it's hard to explain here.

**Comment by teryhill** [comment]
The call_required_libraries is to replace the way assets are currently being called. This needs to be done thru out the code base.

**Comment by aethelwulffe** [comment]
Good idea Miguel.

The call_required_libraries is a fairly new function.  We have been 
migrating the system to use that vs. a ton of js scripts on every page.

The refactor process got stalled several times as we changed the 
function itself.  Lots of people took a stab at it.  We went from bool 
values to the current scheme.

Here is how the process works:

Find an include script in a file.  Copy the dependency from 
/libraries/js (or wherever) to /assets/js/name_of_library/index.js (or 
whatever is appropriate in the /assets directory).

Replace all the js asset calls in the file.  Test.

Find all other files that call the same assets.  Save this list.

Identify the files related to the FEATURE you are working on.

Refactor the calls in all FEATURE files.

Test.

Submit a PR for the new function call for that feature.

NEXT:

Identify another feature that includes any files calling the .js 
dependency from libraries/js.

Repeat the above, submit PR.

When no files are left that call the dependency from the old location, 
submit a PR that DELETES the original.

**Comment by aethelwulffe** [comment]
Please review the calls at the bottom of the file for textformat and dialog.js.
Refactoring to put in conditionals will likely be a very big problem, and outside the scope of this PR, but someone screwed up bigtime by adding them in there. Every file that calls this function will need to be checked for actual usage of textformat.js and dialog.js calls before the function arguments to include them can be implemented. 

This should be done before we actually convert any more files to use this header function.

**Comment by teryhill** [comment]
@aethelwulffe  testformat is used to format the phone numbers , text items etc. Dialog does functions for cascading windows grabbing focus etc. These were done by Rod and are used in most every place that jquery is used. I don't believe they will hurt by being called. There should be a project to check which ones are really used and get rid of what is not.

I am the one who "Screwed up"

**Comment by aethelwulffe** [comment]
No, you are not "the" one.  Every one of us have taken turns "developing" (means screwing up, refactoring again) these functions.  We seem to be headed towards getting it right finally.

**Comment by MigDinny** [comment]
@teryhill @aethelwulffe I don't really know what is the best option. If the intention is to **always** include those scripts, that should work but only when **headers.inc.php** is included also.

Should I let textformat and dialog be there?

**Comment by aethelwulffe** [comment]
Yeah @MigDinny leave them alone.  If we ONLY want those, we call the function with no arguments...which is not a good consistent practice.  That problem is not in the scope of your PR though.

**Comment by MigDinny** [comment]
@aethelwulffe Yep! Maybe I should move those scripts into _call_required_libraries_ instead of letting them at the end of **headers.inc.php**. That way those scripts would only be included if _call_required_libraries_ was called and not if **headers.inc.php** was included. (look closely at the **headers.inc.php** code and you'll get what i mean)

**Comment by MigDinny** [comment]
@aethelwulffe @teryhill 

This is being worked in every PR that has "library refactor" in its title.

I will link them here

**Comment by MigDinny** [comment]
@aethelwulffe @teryhill I wonder if the removal of "html_header_show" is needed. The code I'm touching generally is after that line and it's really easy to me to remove it anyway. Should I do it?

**Comment by yehster** [comment]
My suggestion is that you try it out with and without "html_header_show" and report on your findings...

Run the page and do "view source" in each of the two configurations and see what (if any differences) there are...

**Comment by MigDinny** [comment]
@yehster I did it and nothing happens. That function is only defined in only 1 place and it has nothing inside. It literally does nothing. I was wondering if I should let it be there or remove it.

**Comment by yehster** [comment]
Given the comment for htm_header_show which comes from translations.inc

    // Below line was commented by the UTF-8 project on 05-2009 by BM.
    //  We commented this out since we are now standardizing encoding
    //  in the globals.php file.
    // echo '<meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/> '."\n";

@MigDinny What is YOUR judgement about how to handle html_header_show?

**Comment by MigDinny** [comment]
@yehster 

My judgement would be to **not remove** the function in the codebase (and its callings) for 2 reasons:

- If we need to put something in the headers in every file one day, the function was already created and was already being called in most of the files and would save a lot of _possible_ work;

- The function has no code inside it yet, that means that the resources and time the server uses for those callings are insignificant, therefore, we wouldn't actually gain nothing;

I vote for letting the function stay in the codebase. It can be needed and it is not affecting performance or development, in my opinion.
